src/common/config_validator.cpp

```diff
@@ -0,0 +1,142 @@
+#include "common/config_validator.h"
+#include "3rd_party/exception.h"
+#include "common/logging.h"
+#include "common/regex.h"
+#include "common/utils.h"
+
+#include <boost/filesystem.hpp>
+
+namespace marian {
+
+bool ConfigValidator::has(const std::string& key) const {
+  return config_[key];
+}
+
+ConfigValidator::ConfigValidator(const YAML::Node& config)
+    : config_(config) {}
+
+ConfigValidator::~ConfigValidator() {}
+
+void ConfigValidator::validateOptions(cli::mode mode) const {
+  switch(mode) {
+    case cli::mode::translation:
+      validateOptionsTranslation();
+      break;
+    case cli::mode::scoring:
+      validateOptionsParallelData();
+      validateOptionsScoring();
+      break;
+    case cli::mode::training:
+      validateOptionsParallelData();
+      validateOptionsTraining();
+      break;
+  }
+}
+
+void ConfigValidator::validateOptionsTranslation() const {
```

Comment:
What is not fully clear to me is the different levels. There are classes for
 * parsing cmd-line options
 * representing a set of options
 * interpreting these options in the context of Marian

Which class does what, which class knows what? it looks like ConfigValidator knows about Marian. How about Config? How about Options? Should we qualify classes that know about Marian by including it in the name? E.g. MarianConfig?

Reply:
As said above. This is currently chaotic, but this PR is not meant to fix this yet.

Reply:
Added TODO